desc/utils.py

```diff
@@ -684,4 +697,515 @@ def broadcast_tree(tree_in, tree_out, dtype=int):
+
+    return _f
+
+
+@_treeify
+def _unchunk(x):
+    return x.reshape((-1,) + x.shape[2:])
+
+
+@_treeify
+def _chunk(x, chunk_size=None):
+    # chunk_size=None -> add just a dummy chunk dimension, same as np.expand_dims(x, 0)
+    if x.ndim == 0:
+        raise ValueError("x cannot be chunked as it has 0 dimensions.")
+    n = x.shape[0]
+    if chunk_size is None:
+        chunk_size = n
+
+    n_chunks, residual = divmod(n, chunk_size)
+    if residual != 0:
+        raise ValueError(
+            "The first dimension of x must be divisible by chunk_size."
+            + f"\n            Got x.shape={x.shape} but chunk_size={chunk_size}."
+        )
+    return x.reshape((n_chunks, chunk_size) + x.shape[1:])
+
+
+def _chunk_size(x):
+    b = set(map(lambda x: x.shape[:2], jax.tree_util.tree_leaves(x)))
+    if len(b) != 1:
+        raise ValueError(
+            "The arrays in x have inconsistent chunk_size or number of chunks"
+        )
+    return b.pop()[1]
+
+
+def unchunk(x_chunked):
+    """Merge the first two axes of an array (or a pytree of arrays).
+
+    Parameters
+    ----------
+    x_chunked: an array (or pytree of arrays) of at least 2 dimensions
+
+    Returns
+    -------
+    (x, chunk_fn) : tuple
+        where x is x_chunked reshaped to (-1,)+x.shape[2:]
+        and chunk_fn is a function which restores x given x_chunked
+
+    """
+    return _unchunk(x_chunked), functools.partial(
+        _chunk, chunk_size=_chunk_size(x_chunked)
+    )
+
+
+def chunk(x, chunk_size=None):
+    """Split an array (or a pytree of arrays) into chunks along the first axis.
+
+    Parameters
+    ----------
+        x: an array (or pytree of arrays)
+        chunk_size: an integer or None (default)
+            The first axis in x must be a multiple of chunk_size
+
+    Returns
+    -------
+    (x_chunked, unchunk_fn): tuple
+        - x_chunked is x reshaped to (-1, chunk_size)+x.shape[1:]
+          if chunk_size is None then it defaults to x.shape[0], i.e. just one chunk
+        - unchunk_fn is a function which restores x given x_chunked
+    """
+    return _chunk(x, chunk_size), _unchunk
+
+
+####
+
+# The following section of this code is derived from the NetKet project
+# https://github.com/netket/netket/blob/9881c9fb217a2ac4dc9274a054bf6e6a2993c519/
+# netket/jax/_scanmap.py
+#
+# The original copyright notice is as follows
+# Copyright 2021 The NetKet Authors - All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+
+_tree_add = functools.partial(jax.tree_util.tree_map, jax.lax.add)
+_tree_zeros_like = functools.partial(
+    jax.tree_util.tree_map, lambda x: jnp.zeros(x.shape, dtype=x.dtype)
+)
+
+
+# TODO put it somewhere
+def _multimap(f, *args):
+    try:
+        return tuple(map(lambda a: f(*a), zip(*args)))
+    except TypeError:
+        return f(*args)
+
+
+def scan_append_reduce(f, x, append_cond, op=_tree_add, zero_fun=_tree_zeros_like):
```

Comment:
I think for the use case we care about (`append_cond=True`) this is equivalent to regular `jax.lax.scan`?

Reply:
Possibly? I pared down as much as I could but there is still some logic that is opaque to me preventing me from just making it a bare scan, so I kept `scan_append`